File: agent-service/src/agents/coordinator/pool.ts
import { randomUUID } from "crypto";
import type { CodingAgentDriver } from "./drivers/base";
import { LangGraphDriver } from "./drivers/langgraph";
import { OpenCodeDriver } from "./drivers/opencode";
import { GitHubCopilotDriver } from "./drivers/githubCopilot";
import { BaseAgentDriver } from "./drivers/base-agent-driver";
import { AgentRegistry } from "../../core/agent-framework/registry";
import { ROLE_TO_AGENT_TYPE } from "../register";
import { buildChatModel } from "../../core/llm";
import { decrypt } from "../../core/crypto";
import type { Skill, McpServer } from "../../core/types";

export interface AgentProfileConfig {
  id: string;
  agent_type: "langgraph" | "opencode" | "github-copilot";
  llm_provider: string;
  llm_model: string;
  llm_api_key: string;
  llm_base_url?: string;
  github_token?: string;
  mcp_servers: McpServer[];
  skills: Skill[];
}

export interface ProjectAgentConfig {
  id: string;
  agent_profile_id: string;
  role: string;
  scope: Record<string, unknown> | null;
}

export class AgentPool {
  /** Drivers keyed by agent profile ID (legacy CodingAgentDriver path). */
  private drivers = new Map<string, CodingAgentDriver>();

  /**
   * Drivers keyed by role string (reviewer/tester/knowledge/infra).
   * These wrap BaseAgent instances via BaseAgentDriver.
   */
  private roleDrivers = new Map<string, CodingAgentDriver>();

  initialize(profiles: AgentProfileConfig[]): void {
    for (const profile of profiles) {
      const driver = this.buildDriver(profile);
      this.drivers.set(profile.id, driver);
    }
  }

  /**
   * Create BaseAgentDriver-wrapped drivers for specialist roles
   * (reviewer, tester, knowledge, infra) and index them by role.
   * Must be called after `registerAllAgents()`.
   */
  registerRoleDrivers(
    projectAgents: ProjectAgentConfig[],
    projectCtx: { id: string; name: string },
  ): void {
    const registry = AgentRegistry.getInstance();

    for (const pa of projectAgents) {
      const agentType = ROLE_TO_AGENT_TYPE[pa.role];
      if (!agentType) continue; // planner/coder use legacy CodingAgentDriver path
      if (!registry.has(agentType)) continue;

      const instanceId = `${pa.role}-${randomUUID().slice(0, 8)}`;
      const agent = registry.create(agentType, instanceId);
      const driver = new BaseAgentDriver(agent, projectCtx);

      this.roleDrivers.set(pa.role, driver);
      // Also index by profile ID so profile-based fallback still works
      this.drivers.set(pa.agent_profile_id, driver);
    }
  }

  getDriver(profileId: string): CodingAgentDriver | undefined {
    return this.drivers.get(profileId);
  }

  /**
   * Return the driver registered for a given role (e.g. "reviewer", "tester").
   * Returns undefined if no role driver was registered (caller should fall back
   * to a profile-based driver).
   */
  getDriverByRole(role: string): CodingAgentDriver | undefined {
    return this.roleDrivers.get(role);
  }

  private buildDriver(profile: AgentProfileConfig): CodingAgentDriver {
    const apiKey = decrypt(profile.llm_api_key);
    const githubToken = profile.github_token ? decrypt(profile.github_token) : undefined;

    switch (profile.agent_type) {
      case "opencode":
        return new OpenCodeDriver({
          llmProvider: profile.llm_provider,
          llmModel: profile.llm_model,
          llmApiKey: apiKey,
          llmBaseUrl: profile.llm_base_url,
          skills: profile.skills,
          mcpServers: profile.mcp_servers,
        });

      case "github-copilot":
        return new GitHubCopilotDriver({
          githubToken,
          llmProvider: profile.llm_provider,
          llmApiKey: apiKey,
          llmBaseUrl: profile.llm_base_url,
          skills: profile.skills,
        });

      case "langgraph":
      default: {
        const llm = buildChatModel({
          provider: profile.llm_provider,
          model: profile.llm_model,
          apiKey,
          baseUrl: profile.llm_base_url,
        });
        return new LangGraphDriver({ llm, skills: profile.skills });
      }
    }
  }
}

File: agent-service-py/src/agent_service/agents/coordinator/pool.py
from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from typing import Dict, List, Optional

from agent_service.agents.coordinator.drivers.base import CodingAgentDriver
from agent_service.agents.coordinator.drivers.langgraph import LangGraphDriver
from agent_service.agents.coordinator.drivers.opencode import OpenCodeDriver
from agent_service.agents.coordinator.drivers.github_copilot import GitHubCopilotDriver
from agent_service.agents.coordinator.drivers.base_agent_driver import BaseAgentDriver
from agent_service.agents.register import ROLE_TO_AGENT_TYPE
from agent_service.core.agent_framework.registry import AgentRegistry
from agent_service.core.llm import build_chat_model
from agent_service.core.types import McpServer, Skill
from agent_service.crypto import decrypt


@dataclass
class AgentProfileConfig:
    id: str
    agent_type: str  # "langgraph" | "opencode" | "github-copilot"
    llm_provider: str
    llm_model: str
    llm_api_key: str
    llm_base_url: Optional[str] = None
    github_token: Optional[str] = None
    mcp_servers: List[McpServer] = field(default_factory=list)
    skills: List[Skill] = field(default_factory=list)


class AgentPool:
    """Manages a pool of ``CodingAgentDriver`` instances keyed by profile ID or role."""

    def __init__(self) -> None:
        # Drivers keyed by agent profile ID (LangGraph / OpenCode / Copilot path)
        self._drivers: Dict[str, CodingAgentDriver] = {}
        # Drivers keyed by role string (reviewer / tester / knowledge / infra)
        self._role_drivers: Dict[str, CodingAgentDriver] = {}

    def initialize(self, profiles: List[Dict]) -> None:
        for profile in profiles:
            api_key = decrypt(profile.get("llm_api_key", ""))
            github_token_raw = profile.get("github_token")
            github_token = decrypt(github_token_raw) if github_token_raw else None

            mcp_servers = [McpServer(**s) for s in (profile.get("mcp_servers") or [])]
            skills = [Skill(**s) for s in (profile.get("skills") or [])]

            p = AgentProfileConfig(
                id=profile["id"],
                agent_type=profile.get("agent_type", "langgraph"),
                llm_provider=profile.get("llm_provider", "openai"),
                llm_model=profile.get("llm_model", "gpt-4o"),
                llm_api_key=api_key,
                llm_base_url=profile.get("llm_base_url"),
                github_token=github_token,
                mcp_servers=mcp_servers,
                skills=skills,
            )
            driver = self._build_driver(p)
            self._drivers[p.id] = driver

    def register_role_drivers(
        self,
        project_agents: List[Dict],
        project_ctx: Dict,
    ) -> None:
        """
        Create BaseAgentDriver-wrapped drivers for specialist roles and index them by role.
        Must be called after ``register_all_agents()``.
        """
        registry = AgentRegistry.get_instance()

        for pa in project_agents:
            role = pa.get("role", "")
            agent_type = ROLE_TO_AGENT_TYPE.get(role)
            if not agent_type:
                continue
            if not registry.has(agent_type):
                continue

            instance_id = f"{role}-{str(uuid.uuid4())[:8]}"
            agent = registry.create(agent_type, instance_id)
            driver = BaseAgentDriver(agent, project_ctx)
            self._role_drivers[role] = driver
            self._drivers[pa.get("agent_profile_id", "")] = driver

    def get_driver(self, profile_id: str) -> Optional[CodingAgentDriver]:
        return self._drivers.get(profile_id)

    def get_driver_by_role(self, role: str) -> Optional[CodingAgentDriver]:
        return self._role_drivers.get(role)

    def _build_driver(self, profile: AgentProfileConfig) -> CodingAgentDriver:
        if profile.agent_type == "opencode":
            return OpenCodeDriver(
                llm_provider=profile.llm_provider,
                llm_model=profile.llm_model,
                llm_api_key=profile.llm_api_key,
                llm_base_url=profile.llm_base_url,
                skills=profile.skills,
                mcp_servers=profile.mcp_servers,
            )

        if profile.agent_type == "github-copilot":
            return GitHubCopilotDriver(
                github_token=profile.github_token,
                llm_provider=profile.llm_provider,
                llm_api_key=profile.llm_api_key,
                llm_base_url=profile.llm_base_url,
                skills=profile.skills,
            )

        # LangGraph (default)
        llm = build_chat_model(
            provider=profile.llm_provider,
            model=profile.llm_model,
            api_key=profile.llm_api_key,
            base_url=profile.llm_base_url,
        )
        return LangGraphDriver(llm=llm, skills=profile.skills)
